src/pending.js

```diff
@@ -0,0 +1,128 @@
+var _ = require('lodash');
+var debug = require('debug')('stats-collector:pending');
+var taskcluster = require('taskcluster-client');
+
+const FLUSH_PERIOD = 30; // seconds
+
+// check for pending tasks that aren't really pending with this frequency; should
+// be smaller than MIN_CHECK_AGE
+const CHECK_INTERVAL = 120; // seconds
+
+// only check tasks pending at least this long; this should be lower than any alert
+// thresholds for pending tasks.
+const MIN_CHECK_AGE = 300; // seconds
+
+// calculate the taskKey and scheduled date with the earliest scheduled time for
+// the given workerType; if there are none, returns null and the current time
+var earliest = (state, workerType, now) => {
+  return _.reduce(state.pendingTasks[workerType], (res, scheduled, taskKey) => {
+    if (scheduled < res.scheduled) {
+      return {scheduled, taskKey};
+    } else {
+      return res;
+    }
+  }, {taskKey: null, scheduled: now});
+};
+
+// update the state based on a task message
+var update = (state, workerType, taskKey, isPending, scheduled) => {
+  let workerTypeState = state.pendingTasks[workerType];
+
+  if (!workerTypeState) {
+    state.pendingTasks[workerType] = workerTypeState = {};
+  }
+
+  if (isPending) {
+    debug('task pending: %s at %s (%s)', taskKey, scheduled, workerType);
+    workerTypeState[taskKey] = new Date(scheduled).getTime();
+  } else {
+    debug('task no longer pending: %s (%s)', taskKey, workerType);
+    if (workerTypeState[taskKey]) {
+      state.readyWorkerTypes[workerType] = true;
+      delete workerTypeState[taskKey];
+    }
+  }
+};
+
+// update monitors based on the current state
+var flush = (state, monitor, now) => {
+  for (let workerType of Object.keys(state.readyWorkerTypes)) {
+    let {taskKey, scheduled} = earliest(state, workerType, now);
+    let waiting = taskKey ? now - scheduled : 0;
+    monitor.measure(`tasks.${workerType}.pending`, waiting);
+  }
+};
+
+// For each worker type with a task we think is pending for over MIN_CHECK_AGE,
+// poll the status of that task and, if it is actually not pending, remove it
+// from the list.  This provides a way to catch cases where the service misses
+// a pulse message or they are delivered out of order.
+var check = (state, queue, now) => {
+  return Promise.all(Object.keys(state.pendingTasks).map(async (workerType) => {
+    let {taskKey, scheduled} = earliest(state, workerType, now);
+    if (taskKey && now - scheduled > MIN_CHECK_AGE * 1000) {
+      let [taskId, runId] = taskKey.split('/');
+      let taskStatus = await queue.status(taskId);
+      if (taskStatus.status.runs[runId].state !== 'pending') {
+        debug('task not actually pending: %s (%s)', taskKey, workerType);
+        delete state.pendingTasks[workerType][taskKey];
+      }
```

Comment:
Ah, I had omitted this so as to rate-limit access to the queue -- so if there are several phantom pendings, one would be stripped at each run of check().  But you're familiar with queue's limits, so I'll adopt your suggestion :)